### PR TITLE
stat: Add counterFromStatName(), gaugeFromStatName(), and histogramFromStatName()

### PR DESCRIPTION
**Pull request by jmarantz**
Description: #6161 is a 45-file PR that converts the stats infrastructure to use a SymbolTable API for representing names, which is challenging to review due to several concepts occurring at once. This PR factors out one of the concepts: the new APIs for Stats::Store and Stats::Scope.
Risk Level: low -- this introduces new APIs that are not actually used yet, though they are tested.
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/stats/isolated_store_impl.h [hunk 1]
```diff
@@ -81,11 +79,12 @@ class IsolatedStoreImpl : public Store {
     return std::vector<ParentHistogramSharedPtr>{};
   }
 
+  void clear();
```
Comment (ambuc) [inline source/common/stats/isolated_store_impl.h]: Does this only get called in tests? If so, rename it something test-appropriate (or use a test peer? not sure what best practice is). Else, please provide docs -- does this get called at shutdown?

File: source/common/stats/scope_prefixer.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  // Scope
+  ScopePtr createScope(const std::string& name) override;
+  Counter& counter(const std::string& name) override { return scope_->counter(prefix_ + name); }
+  Gauge& gauge(const std::string& name) override { return scope_->gauge(prefix_ + name); }
+  Histogram& histogram(const std::string& name) override {
+    return scope_->histogram(prefix_ + name);
+  }
+  void deliverHistogramToSinks(const Histogram& histograms, uint64_t val) override;
+  Counter& counterFromStatName(StatName name) override;
+  //    return counter(symbolTable().toString(name));
```
Comment (ambuc) [inline source/common/stats/scope_prefixer.h]: Intentional?

File: source/common/stats/scope_prefixer.cc [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    : prefix_(Utility::sanitizeStatsName(prefix)), scope_(&scope), owns_scope_(false) {}
+
+ScopePrefixer::~ScopePrefixer() {
+  if (owns_scope_) {
+    delete scope_;
+  }
+}
+
+ScopePtr ScopePrefixer::createScope(const std::string& name) {
+  // StatNameStorage scope_stat_name(name, symbolTable());  // Takes a lock.
```
Comment (ambuc) [inline source/common/stats/scope_prefixer.cc]: Intentional?

File: source/common/stats/isolated_store_impl.h (hunk 1 above)
Comment (jmarantz) [reply source/common/stats/isolated_store_impl.h]: I definitely don't need it in this PR and I'm not sure I really need it in future ones either. Removed.

**Comment by jmarantz** [comment]
I'm a little reluctant to hack the doc as this is not a meaningful change to the operation of stats in this PR, just a new API this is defined and tested but not used. #6161 changes operations and explains how and why in stats.md.

**Comment by jmarantz** [comment]
Matt -- this should be pretty easy; an incremental step requested by @ambuc in advance of #6161

File: source/common/stats/scope_prefixer.cc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+namespace Stats {
+
+// Variant of ScopePrefixer that owns the scope being prefixed, and will
+// delete it upon destruction.
+ScopePrefixer::ScopePrefixer(absl::string_view prefix, Scope* scope)
+    : prefix_(Utility::sanitizeStatsName(prefix)), scope_(scope), owns_scope_(true) {}
+
+// Variant of ScopePrefixer that references the scope being prefixed, but does
+// not own it.
+ScopePrefixer::ScopePrefixer(absl::string_view prefix, Scope& scope)
```
Comment (mattklein123) [inline source/common/stats/scope_prefixer.cc]: I don't see this variant used, is it used in a follow up?

Also, somewhat nitty, but for clarity I would prefer a single constructor that looks like:
```
ScopePrefixer::ScopePrefixer(absl::string_view prefix, Scope& scope, bool take_ownership)
```

And then personally I would have the following members:
```
std::unique_ptr<Scope> owned_scope_;
Scope* scope_to_use_;
```
And avoid the manual deletion. WDYT?

Comment (jmarantz) [reply source/common/stats/scope_prefixer.cc]: Yeah I like your pattern better. However I also can't justify the speculative generality here; I needed it in some past iteration and don't see a user-case now. If I have to add it back in the future I'll use your pattern.

update coming.

**Pull request merged**